### ISSUE TITLE
Core/Creature: NPC positioning during combat [$25 bounty]

### ISSUE THREAD

**Issue by Teppic1** [TrinityCore/TrinityCore#20922]
**Description:**

Two prominent combat features from retail are currently missing in TrinityCore:

1. If a single NPC is too close to the player it should reposition itself by walking backwards.
2. If multiple NPCs are attacking the same player, they should reposition themselves after a brief period by walking to form a fan/cone shape in front of the player so they do not stack on top of each other.

I don't have the exact distances and timings for the delays but they should be easy to get from sniffs and videos.

These problems are fixed variously by some private servers implementing TrinityCore so I imagine aren't a massive undertaking (e.g. Rising Gods has the 'fan' mechanic, and that's based on an old version of TC).

This encompasses bug #20778 but is part of a bigger combat positioning issue.

**Branch(es):**

both

**TC rev. hash/commit:** 

TrinityCore rev. 59a4ea1edce3+

**Comment by nubqt** [comment]
The second expected behaviour is incorrect: the blizzlike way is for the NPCs to form a circle, not a cone. You can test this behavior in Blackwing Lair with the Nefarian encounter. Phase one of that encounter spawns an infinite number of drakonids until you kill 42 of them. If you don't kill any and just stand still, they will all eventually make a circle around you. You can make a level 110 character to test this behavior since the drakonids will barely damage you.

If I recall correctly, the circle forming takes two considerations:
1. From where the NPC rushed you
2. Whether there is an NPC already hitting you in that spot

So, if you have an NPC rushing you from your front, he will try to settle himself as close as possible to that spot where he should be attacking you from. If there is an NPC there, he goes either right or left of 'his spot.'

But if he rushed you from your back, he will settle himself at your back (not move to your front). And perform the same logic as described above (move left or right if there is an NPC already at 'his spot').

**Comment by Teppic1** [comment]
On the older report for mobs moving backwards a couple of people said that they shouldn't encircle but should form a cone. It's possible that this changed between Wrath and Legion (it'd be easy to verify in Legion though, just go into a low level instance at max level and pull everything).

**Comment by offl** [comment]
> 1. If a single NPC is too close to the player it should reposition itself by walking backwards.

#20778

> 2. If multiple NPCs are attacking the same player, they should reposition themselves after a brief period by walking to form a fan/cone shape in front of the player so they do not stack on top of each other.

18512

**Comment by Teppic1** [comment]
Note a single issue is required to assign a bounty:

https://www.bountysource.com/issues/51685422-core-creature-npc-positioning-during-combat-25-bounty

**Comment by Langerz82** [comment]
test patch:
https://gist.github.com/Langerz82/7242c367e5d533f66ef3720b8680de62

**Comment by offl** [comment]
What about https://github.com/TrinityCore/TrinityCore/pull/21282?

**Comment by Langerz82** [comment]
@offl the patch can be combined with the proposed changes #21282 does.

**Comment by miraco** [comment]
Did this got fixed already?

**Comment by tkrokli** [comment]
I don't think so, haven't noticed that NPC creatures fan out if multiples attack player on TC compared to retail.
Do you have any indications of improved AI behaviour on TC related to this issue?

**Comment by Langerz82** [comment]
I got a patch for this, regarding moving back I created a simple function to do so, until OvahLords more extensive PR is merged.
Will submit patch very soon.

**Comment by Langerz82** [comment]
See:
https://gist.github.com/Langerz82/2ba589313bff0880e9217a9a49238d8b

**Comment by Langerz82** [comment]
Ready for testing:
https://gist.github.com/Langerz82/2ba589313bff0880e9217a9a49238d8b

It doesnt fan the mobs out perfectly the only try to fan when chasing not while in attack mode. That would add an extra layer of complexity I cant be stuffed programming atm.

**Comment by Langerz82** [comment]
Was gonna wait till re-doing this, then realised I dont have a life. lol
Anyway here is the new patch for testing:
https://gist.github.com/Langerz82/1dbe4acec7fb6bbedf61fb4a44aeece3

**Comment by offl** [comment]
More info #18512

**Comment by Teppic1** [comment]
Bounty was removed due to bountysource being extremely scummy the other month.

**Comment by offl** [comment]
Moreover creature can predict player's destination and stop moving if player moves forwards to creature https://youtu.be/c9DT8esj9Aw?t=57

**Comment by CraftedRO** [comment]
33118dabd28bb3a7101d481641a5d13ea2006e01 still needed.
